### ISSUE TITLE
Improve smart indenting 

### ISSUE THREAD

**Issue by cmr** [PistonDevelopers/VisualRust#11]
- [x] Make it token based, not textual (may require some infrastructure work to expose lexing of lines)
- [x] Don't look as far back for opening `{` (only keep looking if prevLines are *only* whitespace)
- [ ] Handle things like breaking function calls on multiple lines (should align with first argument)
- [X] ~~Dedent on `}`~~

**Comment by xilec** [comment]
I suppose that all is much easie and harder at same time.
I just examine how it works in Visual Studio (for C#) + Resharper.

It looks like that for true smart indention we should count indention not from plain text, but from abstract syntax tree (AST) builded for current editing document. 

In this way we would identify a syntax block (struct declaration, function body declaration, if statement and so on) in which situated the caret and count correct indention for current case. It should be noted that we connecting not with boundaries of "brackets body", but with boundaries of syntax block. In most cases the indention will be calculated as first simbol of this block position + indention step, but for some cases most suitable is another variants. For instance, for *if statement* with multiline conditions should be better variant as first block position + 3 (offset "if "):
```
............if 
...............|.{
```
where | is caret position.

But nowadays, for improving current smart indentions we can just count different variants of opening and closing brackets and calculate position as indention step + first position of line where current "bracket body" is begining.

**Comment by vosen** [comment]
I've dome some more checking with Visual Studio and turns out that there is one mode for for syntactically correct code and one for syntactically incorrect code.
In case of no syntax errors VS indents to the start of syntactic construct that opened the block.
For example
```
····················static
········void·Main
·············(string[]·args)·{
```
indents to 
```
····················static
········void·Main
·············(string[]·args)·{
························}
```

So actually
```
··································static·Main()·{
```
indents to
```
··································static·Main()·{
······································}
```


but
```
static·······························Main()·{
····}
```
indents to
```
static·······························Main()·{
····}
```
Same about `(` indenting.
In case of syntax errors VS falls back to simpler algorithm I've described above.

**Comment by vosen** [comment]
Closed by #88, further work moved to #93 and #94.
